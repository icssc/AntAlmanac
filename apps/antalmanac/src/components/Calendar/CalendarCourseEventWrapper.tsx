import { Box } from '@mui/material';
import { useCallback, useEffect, useRef } from 'react';
import { EventWrapperProps } from 'react-big-calendar';
import { shallow } from 'zustand/shallow';

import type { CalendarEvent } from '$components/Calendar/CourseCalendarEvent';
import { useSelectedEventStore } from '$stores/SelectedEventStore';

interface CalendarCourseEventWrapperProps extends EventWrapperProps<CalendarEvent> {
    children?: React.ReactNode;
}

/**
 * CalendarCourseEventWrapper allows us to override the default onClick event behavior which problamtically rerenders the entire calendar
 */
export const CalendarCourseEventWrapper = ({ children, ...props }: CalendarCourseEventWrapperProps) => {
    const ref = useRef<HTMLDivElement>(null);

    const setSelectedEvent = useSelectedEventStore((state) => state.setSelectedEvent, shallow);

    const handleClick = useCallback(
        (e: React.MouseEvent) => {
            e.preventDefault();
            e.stopPropagation();

            setSelectedEvent(e, props.event);
        },
        [props.event, setSelectedEvent]
    );

    useEffect(() => {
        const node = ref.current;
        if (!node) {
            return;
        }

        const rbcEvent = node.querySelector('.rbc-event') as HTMLDivElement;
        if (!rbcEvent) {
            return;
        }

        rbcEvent.onclick = (e) => handleClick(e as unknown as React.MouseEvent); // the native onclick requires a little type hacking
    }, [handleClick]);

    return <Box ref={ref}>{children}</Box>;
};
