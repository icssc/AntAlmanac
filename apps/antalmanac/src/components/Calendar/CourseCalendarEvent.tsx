import { Link } from 'react-router-dom';
import { Chip, IconButton, Paper, Tooltip } from '@material-ui/core';
import { Theme, withStyles } from '@material-ui/core/styles';
import { ClassNameMap, Styles } from '@material-ui/core/styles/withStyles';
import { Delete } from '@material-ui/icons';
import { Event } from 'react-big-calendar';
import { useEffect, useRef, useCallback } from 'react';

import CustomEventDialog from './Toolbar/CustomEventDialog/CustomEventDialog';
import { deleteCourse, deleteCustomEvent } from '$actions/AppStoreActions';
import ColorPicker from '$components/ColorPicker';
import analyticsEnum, { logAnalytics } from '$lib/analytics';
import { clickToCopy, isDarkMode } from '$lib/helpers';
import AppStore from '$stores/AppStore';
import locationIds from '$lib/location_ids';
import { useTabStore } from '$stores/TabStore';
import { formatTimes } from '$stores/calendarizeHelpers';
import { useTimeFormatStore } from '$stores/TimeStore';
import buildingCatalogue from '$lib/buildingCatalogue';

const styles: Styles<Theme, object> = {
    courseContainer: {
        padding: '0.5rem',
        margin: '0 1rem',
        minWidth: '15rem',
    },
    customEventContainer: {
        padding: '0.5rem',
    },
    buttonBar: {
        display: 'flex',
        alignItems: 'center',
    },
    title: {
        fontSize: '0.9rem',
        fontWeight: 500,
    },
    icon: {
        cursor: 'pointer',
    },
    titleBar: {
        display: 'flex',
        justifyContent: 'space-between',
        alignItems: 'center',
        marginBottom: '0.25rem',
    },
    table: {
        border: 'none',
        width: '100%',
        borderCollapse: 'collapse',
        fontSize: '0.9rem',
    },
    alignToTop: {
        verticalAlign: 'top',
    },
    rightCells: {
        textAlign: 'right',
    },
    multiline: {
        whiteSpace: 'pre',
    },
    stickToRight: {
        float: 'right',
    },
    colorPicker: {
        cursor: 'pointer',
        '& > div': {
            margin: '0px 8px 0px 4px',
            height: '20px',
            width: '20px',
            borderRadius: '50%',
        },
    },

    clickableLocation: {
        cursor: 'pointer',
        color: isDarkMode() ? '#1cbeff' : 'blue',
        background: 'none !important',
        border: 'none',
        padding: '0 !important',
        fontSize: 'inherit',
        textDecoration: 'none',
    },
};

interface CommonCalendarEvent extends Event {
    color: string;
    start: Date;
    end: Date;
    title: string;
}

export interface Location {
    /**
     * @example 'ICS'
     */
    building: string;

    /**
     * @example '174'
     */
    room: string;

    /**
     * If the location only applies on specific days, this is non-null.
     */
    days?: string;
}

export interface FinalExam {
    examStatus: 'NO_FINAL' | 'TBA_FINAL' | 'SCHEDULED_FINAL';
    dayOfWeek: 'Sun' | 'Mon' | 'Tue' | 'Wed' | 'Thu' | 'Fri' | 'Sat' | null;
    month: number | null;
    day: number | null;
    startTime: {
        hour: number;
        minute: number;
    } | null;
    endTime: {
        hour: number;
        minute: number;
    } | null;
    locations: Location[] | null;
}

export interface CourseEvent extends CommonCalendarEvent {
    locations: Location[];
    showLocationInfo: boolean;
    finalExam: FinalExam;
    courseTitle: string;
    instructors: string[];
    isCustomEvent: false;
    sectionCode: string;
    sectionType: string;
    term: string;
}

/**
 * There is another CustomEvent interface in CourseCalendarEvent and they are slightly different.  The this one represents only one day, like the event on Monday, and needs to be duplicated to be repeated across multiple days. The other one, `CustomEventDialog`'s `RepeatingCustomEvent`, encapsulates the occurences of an event on multiple days, like Monday Tuesday Wednesday all in the same object as specified by the `days` array.
 * https://github.com/icssc/AntAlmanac/wiki/The-Great-AntAlmanac-TypeScript-Rewritening%E2%84%A2#duplicate-interface-names-%EF%B8%8F
 */
export interface CustomEvent extends CommonCalendarEvent {
    customEventID: number;
    isCustomEvent: true;
    building: string;
    days: string[];
}

export type CalendarEvent = CourseEvent | CustomEvent;

interface CourseCalendarEventProps {
    classes: ClassNameMap;
    courseInMoreInfo: CalendarEvent;
    scheduleNames: string[];
    closePopover: () => void;
}

const MONTHS = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];

const CourseCalendarEvent = (props: CourseCalendarEventProps) => {
    const paperRef = useRef<HTMLInputElement>(null);

    useEffect(() => {
        const handleKeyDown = (event: { keyCode: number }) => {
            // event.keyCode === 27 reads for the "escape" key
            if (event.keyCode === 27) {
                if (paperRef.current) paperRef.current.style.display = 'none';
            }
        };

        document.addEventListener('keydown', handleKeyDown);

        return () => {
            document.removeEventListener('keydown', handleKeyDown);
        };
    }, []);

    const { setActiveTab } = useTabStore();
    const { isMilitaryTime } = useTimeFormatStore();

    const focusMap = useCallback(() => {
        setActiveTab(2);
    }, [setActiveTab]);

    const { classes, courseInMoreInfo } = props;

    if (!courseInMoreInfo.isCustomEvent) {
        const { term, instructors, sectionCode, title, finalExam, locations, sectionType } = courseInMoreInfo;

        let finalExamString = '';

        if (finalExam.examStatus == 'NO_FINAL') {
            finalExamString = 'No Final';
        } else if (finalExam.examStatus == 'TBA_FINAL') {
            finalExamString = 'Final TBA';
        } else {
            if (finalExam.startTime && finalExam.endTime && finalExam.month && finalExam.locations) {
                const timeString = formatTimes(finalExam.startTime, finalExam.endTime, isMilitaryTime);
                const locationString = `at ${finalExam.locations
                    .map((location) => `${location.building} ${location.room}`)
                    .join(', ')}`;
                const finalExamMonth = MONTHS[finalExam.month];

                finalExamString = `${finalExam.dayOfWeek} ${finalExamMonth} ${finalExam.day} ${timeString} ${locationString}`;
            }
        }

        return (
            <Paper className={classes.courseContainer} ref={paperRef}>
                <div className={classes.titleBar}>
                    <span className={classes.title}>{`${title} ${sectionType}`}</span>
                    <Tooltip title="Delete">
                        <IconButton
                            size="small"
                            onClick={() => {
                                deleteCourse(sectionCode, term);
                                logAnalytics({
                                    category: analyticsEnum.calendar.title,
                                    action: analyticsEnum.calendar.actions.DELETE_COURSE,
                                });
                            }}
                        >
                            <Delete fontSize="inherit" />
                        </IconButton>
                    </Tooltip>
                </div>
                <table className={classes.table}>
                    <tbody>
                        <tr>
                            <td className={classes.alignToTop}>Section code</td>
                            <Tooltip title="Click to copy course code" placement="right">
                                <td className={classes.rightCells}>
                                    <Chip
                                        onClick={(event) => {
                                            clickToCopy(event, sectionCode);
                                            logAnalytics({
                                                category: analyticsEnum.classSearch.title,
                                                action: analyticsEnum.classSearch.actions.COPY_COURSE_CODE,
                                            });
                                        }}
                                        className={classes.sectionCode}
                                        label={sectionCode}
                                        size="small"
                                    />
                                </td>
                            </Tooltip>
                        </tr>
                        <tr>
                            <td className={classes.alignToTop}>Term</td>
                            <td className={classes.rightCells}>{term}</td>
                        </tr>
                        <tr>
                            <td className={classes.alignToTop}>Instructors</td>
                            <td className={`${classes.multiline} ${classes.rightCells}`}>{instructors.join('\n')}</td>
                        </tr>
                        <tr>
                            <td className={classes.alignToTop}>Location{locations.length > 1 && 's'}</td>
                            <td className={`${classes.multiline} ${classes.rightCells}`}>
                                {locations.map((location) => (
                                    <div key={`${sectionCode} @ ${location.building} ${location.room}`}>
                                        <Link
                                            className={classes.clickableLocation}
                                            to={`/map?location=${locationIds[location.building] ?? 0}`}
                                            onClick={focusMap}
                                        >
                                            {location.building} {location.room}
                                        </Link>
                                    </div>
                                ))}
                            </td>
                        </tr>
                        <tr>
                            <td>Final</td>
                            <td className={classes.rightCells}>{finalExamString}</td>
                        </tr>
                        <tr>
                            <td>Color</td>
                            <td className={`${classes.colorPicker} ${classes.stickToRight}`}>
                                <ColorPicker
                                    color={courseInMoreInfo.color}
                                    isCustomEvent={courseInMoreInfo.isCustomEvent}
                                    sectionCode={courseInMoreInfo.sectionCode}
                                    term={courseInMoreInfo.term}
                                    analyticsCategory={analyticsEnum.calendar.title}
                                />
                            </td>
                        </tr>
                    </tbody>
                </table>
            </Paper>
        );
    } else {
        const { title, customEventID, building } = courseInMoreInfo;
        return (
            <Paper className={classes.customEventContainer} ref={paperRef}>
                <div className={classes.title}>{title}</div>
                <div className={classes.table}>
                    Location: &nbsp;&nbsp;
                    <Link
                        className={classes.clickableLocation}
                        to={`/map?location=${building ?? 0}`}
                        onClick={focusMap}
                    >
                        {building ? buildingCatalogue[+building].name : ''}
                    </Link>
                </div>
                <div className={classes.buttonBar}>
                    <div className={`${classes.colorPicker}`}>
                        <ColorPicker
                            color={courseInMoreInfo.color}
                            isCustomEvent={true}
                            customEventID={courseInMoreInfo.customEventID}
                            analyticsCategory={analyticsEnum.calendar.title}
                        />
                    </div>
                    <CustomEventDialog
                        onDialogClose={props.closePopover}
                        customEvent={AppStore.schedule.getExistingCustomEvent(customEventID)}
                        scheduleNames={props.scheduleNames}
                    />

                    <Tooltip title="Delete">
                        <IconButton
                            onClick={() => {
                                props.closePopover();
                                deleteCustomEvent(customEventID);
                                logAnalytics({
                                    category: analyticsEnum.calendar.title,
                                    action: analyticsEnum.calendar.actions.DELETE_CUSTOM_EVENT,
                                });
                            }}
                        >
                            <Delete fontSize="small" />
                        </IconButton>
                    </Tooltip>
                </div>
            </Paper>
        );
    }
};

export default withStyles(styles)(CourseCalendarEvent);
