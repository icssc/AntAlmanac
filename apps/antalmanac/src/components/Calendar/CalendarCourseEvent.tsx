import { Box } from '@material-ui/core';
import { memo } from 'react';
import { shallow } from 'zustand/shallow';

import { CalendarEvent } from '$components/Calendar/CourseCalendarEvent';
import locationIds from '$lib/location_ids';
import { useSelectedEventStore } from '$stores/SelectedEventStore';

export const CalendarCourseEvent = memo(({ event }: { event: CalendarEvent }) => {
    const setSelectedEvent = useSelectedEventStore((state) => state.setSelectedEvent, shallow);

    const handleClick = (e: React.MouseEvent) => {
        e.preventDefault();
        e.stopPropagation();

        setSelectedEvent(e, event);
    };

    if (event.isCustomEvent) {
        return (
            <Box onClick={handleClick}>
                <Box
                    style={{
                        display: 'flex',
                        flexWrap: 'wrap',
                        justifyContent: 'space-between',
                        fontWeight: 500,
                        fontSize: '0.8rem',
                    }}
                >
                    <Box>{event.title}</Box>
                </Box>

                <Box style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-between', fontSize: '0.7rem' }}>
                    <Box>{Object.keys(locationIds).find((key) => locationIds[key] === parseInt(event.building))}</Box>
                </Box>
            </Box>
        );
    }

    return (
        <Box onClick={handleClick}>
            <Box
                style={{
                    display: 'flex',
                    flexWrap: 'wrap',
                    justifyContent: 'space-between',
                    fontWeight: 500,
                    fontSize: '0.8rem',
                }}
            >
                <Box>{event.title}</Box>
                <Box style={{ fontSize: '0.8rem' }}> {event.sectionType}</Box>
            </Box>
            <Box style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-between', fontSize: '0.7rem' }}>
                <Box>
                    {event.showLocationInfo
                        ? event.locations.map((location) => `${location.building} ${location.room}`).join(', ')
                        : event.locations.length > 1
                          ? `${event.locations.length} Locations`
                          : `${event.locations[0].building} ${event.locations[0].room}`}
                </Box>
                <Box>{event.sectionCode}</Box>
            </Box>
        </Box>
    );
});

CalendarCourseEvent.displayName = 'CalendarCourseEvent';
