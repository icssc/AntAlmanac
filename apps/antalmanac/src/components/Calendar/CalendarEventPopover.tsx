import { Popover } from '@mui/material';
import { useCallback, useEffect, useState } from 'react';
import { shallow } from 'zustand/shallow';

import CourseCalendarEvent from '$components/Calendar/CourseCalendarEvent';
import AppStore from '$stores/AppStore';
import { useSelectedEventStore } from '$stores/SelectedEventStore';

export function CalendarEventPopover() {
    const [anchorEl, selectedEvent, setSelectedEvent] = useSelectedEventStore(
        (state) => [state.selectedEventAnchorEl, state.selectedEvent, state.setSelectedEvent],
        shallow
    );

    const [scheduleNames, setScheduleNames] = useState(() => AppStore.getScheduleNames());

    const handleClosePopover = useCallback(() => {
        setSelectedEvent(null, null);
    }, [setSelectedEvent]);

    useEffect(() => {
        const updateScheduleNames = () => {
            setScheduleNames(AppStore.getScheduleNames());
        };

        AppStore.on('scheduleNamesChange', updateScheduleNames);

        return () => {
            AppStore.off('scheduleNamesChange', updateScheduleNames);
        };
    }, []);

    if (!selectedEvent) {
        return null;
    }

    return (
        <Popover
            anchorEl={anchorEl}
            open={Boolean(anchorEl) && !!selectedEvent}
            onClose={handleClosePopover}
            anchorOrigin={{
                vertical: 'top',
                horizontal: 'right',
            }}
            transformOrigin={{
                vertical: 'top',
                horizontal: 'left',
            }}
        >
            <CourseCalendarEvent
                closePopover={handleClosePopover}
                selectedEvent={selectedEvent}
                scheduleNames={scheduleNames}
            />
        </Popover>
    );
}
