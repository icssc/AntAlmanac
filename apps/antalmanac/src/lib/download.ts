import type { Course, HourMinute, WebsocSectionFinalExam } from 'peterportal-api-next-types';
import { saveAs } from 'file-saver';
import { createEvents, type EventAttributes } from 'ics';
import { notNull } from './utils';
import { openSnackbar } from '$actions/AppStoreActions';
import analyticsEnum, { logAnalytics } from '$lib/analytics';
import { getDefaultTerm, termData } from '$lib/termData';
import { CourseEvent, CustomEvent, FinalExam } from '$components/Calendar/CourseCalendarEvent';
import AppStore from '$stores/AppStore';

export const quarterStartDates = Object.fromEntries(
    termData
        .filter((term) => term.startDate !== undefined)
        .map((term) => [term.shortName, term.startDate as [number, number, number]])
);

export const months: Record<string, number> = { Mar: 3, Jun: 6, Jul: 7, Aug: 8, Sep: 9, Dec: 12 };

export const daysOfWeek = ['Su', 'M', 'Tu', 'W', 'Th', 'F', 'Sa'] as const;

export const daysOffset: Record<string, number> = { SU: -1, MO: 0, TU: 1, WE: 2, TH: 3, FR: 4, SA: 5 };

export const fallDaysOffset: Record<string, number> = { TH: 0, FR: 1, SA: 2, SU: 3, MO: 4, TU: 5, WE: 6 };

export const translateDaysForIcs = { Su: 'SU', M: 'MO', Tu: 'TU', W: 'WE', Th: 'TH', F: 'FR', Sa: 'SA' };

export const vTimeZoneSection =
    'BEGIN:VTIMEZONE\n' +
    'TZID:America/Los_Angeles\n' +
    'X-LIC-LOCATION:America/Los_Angeles\n' +
    'BEGIN:DAYLIGHT\n' +
    'TZOFFSETFROM:-0800\n' +
    'TZOFFSETTO:-0700\n' +
    'TZNAME:PDT\n' +
    'DTSTART:19700308T020000\n' +
    'RRULE:FREQ=YEARLY;BYMONTH=3;BYDAY=2SU\n' +
    'END:DAYLIGHT\n' +
    'BEGIN:STANDARD\n' +
    'TZOFFSETFROM:-0700\n' +
    'TZOFFSETTO:-0800\n' +
    'TZNAME:PST\n' +
    'DTSTART:19701101T020000\n' +
    'RRULE:FREQ=YEARLY;BYMONTH=11;BYDAY=1SU\n' +
    'END:STANDARD\n' +
    'END:VTIMEZONE\n' +
    'BEGIN:VEVENT';

export const CALENDAR_ID = 'antalmanac/ics';

export const CALENDAR_OUTPUT = 'local' as const;

/**
 * @example [YEAR, MONTH, DAY, HOUR, MINUTE]
 */
export type DateTimeArray = [number, number, number, number, number];

/**
 * @example [YEAR, MONTH, DAY]
 */
export type YearMonthDay = [number, number, number];

/**
 * Get the days that a class occurs.
 * Given a string of days, convert it to a list of days in ics format
 *
 * @example ("TuThF") -> ["TU", "TH", "FR"]
 */
export function getByDays(days: string): string[] {
    return daysOfWeek.filter((day) => days.includes(day)).map((day) => translateDaysForIcs[day]);
}

/**
 * Get the start date of a class
 * Given the term and bydays, this computes the start date of the class.
 *
 * @example ("2021 Spring", 'Tu') -> [2021, 3, 30]
 */
export function getClassStartDate(term: string, bydays: string[]) {
    // Get the start date of the quarter (Monday)
    const quarterStartDate = new Date(...quarterStartDates[term]);

    // The number of days since the start of the quarter.
    let dayOffset: number;

    // Since Fall quarter starts on a Thursday,
    // the first byday and offset will be different from other quarters.
    // Sort by this ordering: [TH, FR, SA, SU, MO, TU, WE]
    if (getQuarter(term) === 'Fall') {
        bydays.sort((day1, day2) => {
            return fallDaysOffset[day1] - fallDaysOffset[day2];
        });
        dayOffset = fallDaysOffset[bydays[0]];
    } else {
        dayOffset = daysOffset[bydays[0]];
    }

    // Add the dayOffset to the quarterStartDate
    // Date object will handle potential overflow into the next month
    quarterStartDate.setDate(quarterStartDate.getDate() + dayOffset);

    // Return [Year, Month, Date]
    return dateToIcs(quarterStartDate);
}

/**
 * Convert a Date object to ics format, i.e. [YYYY, MM, DD]
 */
export function dateToIcs(date: Date) {
    return [
        date.getFullYear(),
        date.getMonth() + 1, // Add 1 month since it is 0-indexed
        date.getDate(),
    ] as YearMonthDay;
}

/**
 * Get the start and end datetime of the first class.
 *
 * @example ([2021, 3, 30], " 4:00-4:50p") -> [[2021, 3, 30, 16, 0], [2021, 3, 30, 16, 50]]
 */
export function getFirstClass(
    date: YearMonthDay,
    startTime: HourMinute,
    endTime: HourMinute
): [DateTimeArray, DateTimeArray] {
    const [classStartTime, classEndTime] = parseTimes(startTime, endTime);
    return [
        [...date, ...classStartTime],
        [...date, ...classEndTime],
    ];
}

/**
 * Get the start and end datetime of an exam
 *
 * @example
 *
 * ```ts
 * const exam = {
 *   month: 6,
 *   day: 7,
 *   startTime: {
 *      hour: 10,
 *      minute: 30,
 *   },
 *   endTime: {
 *     hour: 12,
 *     minute: 30,
 *   },
 *   ...
 * }
 *
 * const year = 2019
 *
 * const [examStart, examEnd] = getExamTime(exam, year)
 *
 * // examStart = [2019, 6, 7, 10, 30]
 * // examEnd = [2019, 6, 7, 12, 30]
 * ```
 */
export function getExamTime(exam: FinalExam, year: number): [DateTimeArray, DateTimeArray] | [] {
    if (exam.month && exam.day && exam.startTime && exam.endTime) {
        const month = exam.month;
        const day = exam.day;
        const [examStartTime, examEndTime] = parseTimes(exam.startTime, exam.endTime);

        return [
            [year, month + 1, day, ...examStartTime],
            [year, month + 1, day, ...examEndTime],
        ];
    } else {
        // This should never happen, but we return an empty array for typescript purposes
        return [];
    }
}

/**
 * Helper to convert a time string to an array format.
 *
 * @example { hour: 16, minute: 0}, { hour: 16, minute: 50 } -> [[16, 0], [16, 50]]
 */
export function parseTimes(startTime: HourMinute, endTime: HourMinute) {
    return [
        [startTime.hour, startTime.minute],
        [endTime.hour, endTime.minute],
    ] as const;
}

/**
 * Get the year of a given term.
 *
 * @example "2019 Fall" -> "2019"
 */
export function getYear(term: string) {
    return parseInt(term.split(' ')[0]);
}

/**
 * Get the quarter of a given term.
 *
 * @example "2019 Fall" -> "Fall"
 */
export function getQuarter(term: string) {
    return term.split(' ')[1];
}

/**
 * Get the number of weeks in a given term.
 *
 * @example 10 for quarters and Summer Session 10wk, 5 for Summer Sessions I and II.
 */
export function getTermLength(quarter: string) {
    return quarter.startsWith('Summer') && quarter !== 'Summer10wk' ? 5 : 10;
}

/**
 * Get a string representing the recurring rule for the VEvent.
 *
 * @example ["TU", "TH"] -> "FREQ=WEEKLY;BYDAY=TU,TH;INTERVAL=1;COUNT=20"
 */
export function getRRule(bydays: string[], quarter: string) {
    /**
     * Number of occurences in the quarter
     */
    let count = getTermLength(quarter) * bydays.length;

    switch (quarter) {
        case 'Fall':
            for (const byday of bydays) {
                switch (byday) {
                    case 'TH':
                    case 'FR':
                    case 'SA':
                        count += 1; // account for Week 0 course meetings
                        break;
                    default:
                        break;
                }
            }
            break;
        case 'Summer1':
            if (bydays.includes('MO')) count += 1; // instruction ends Monday of Week 6
            break;
        case 'Summer10wk':
            if (bydays.includes('FR')) count -= 1; // instruction ends Thursday of Week 10
            break;
        default:
            break;
    }

    return `FREQ=WEEKLY;BYDAY=${bydays.toString()};INTERVAL=1;COUNT=${count}`;
}

export function getEventsFromCourses(events = AppStore.getEventsWithFinalsInCalendar()): EventAttributes[] {
    const calendarEvents = events.flatMap((event) => {
        if (event.isCustomEvent) {
            // FIXME: We don't have a way to get the term for custom events,
            // so we just use the default term.
            const { title, start, end } = event as CustomEvent;
            const days = getByDays(event.days.join(''));
            const term = getDefaultTerm().shortName;
            const rrule = getRRule(days, getQuarter(term));
            const eventStartDate = getClassStartDate(term, days);
            const [firstClassStart, firstClassEnd] = getFirstClass(
                eventStartDate,
                { hour: start.getHours(), minute: start.getMinutes() },
                { hour: end.getHours(), minute: end.getMinutes() }
            );
            const customEvent: EventAttributes = {
                productId: 'antalmanac/ics',
                startOutputType: 'local' as const,
                endOutputType: 'local' as const,
                title: title,
                // TODO: Add location to custom events, waiting for https://github.com/icssc/AntAlmanac/issues/249
                // location: `${location.building} ${location.room}`,
                start: firstClassStart,
                end: firstClassEnd,
                recurrenceRule: rrule,
            };
            return customEvent;
        } else {
            const { term, title, courseTitle, instructors, sectionType, start, end, finalExam } = event;
            const courseEvents: EventAttributes[] = event.locations
                .map((location) => {
                    if (location.days === undefined) {
                        return null;
                    }
                    const days = getByDays(location.days);

                    const [finalStart, finalEnd] = getExamTime(finalExam, getYear(term));

                    if (sectionType === 'Fin') {
                        return {
                            productId: CALENDAR_ID,
                            startOutputType: CALENDAR_OUTPUT,
                            endOutputType: CALENDAR_OUTPUT,
                            title: `${title} Final Exam`,
                            description: `Final Exam for ${courseTitle}`,
                            start: finalStart!,
                            end: finalEnd!,
                        };
                    } else {
                        const classStartDate = getClassStartDate(term, days);

                        const [firstClassStart, firstClassEnd] = getFirstClass(
                            classStartDate,
                            { hour: start.getHours(), minute: start.getMinutes() },
                            { hour: end.getHours(), minute: end.getMinutes() }
                        );

                        const rrule = getRRule(days, getQuarter(term));

                        // Add VEvent to events array.
                        return {
                            productId: 'antalmanac/ics',
                            startOutputType: 'local' as const,
                            endOutputType: 'local' as const,
                            title: `${title} ${sectionType}`,
                            description: `${courseTitle}\nTaught by ${instructors.join('/')}`,
                            location: `${location.building} ${location.room}`,
                            start: firstClassStart,
                            end: firstClassEnd,
                            recurrenceRule: rrule,
                        };
                    }
                })
                .filter(notNull);
            return courseEvents;
        }
    });
    return calendarEvents;
}

export function exportCalendar() {
    const events = getEventsFromCourses();

    // Convert the events into a vcalendar.
    // Callback function triggers a download of the .ics file
    createEvents(events, (error, value) => {
        logAnalytics({
            category: 'Calendar Pane',
            action: analyticsEnum.calendar.actions.DOWNLOAD,
        });

        if (error) {
            openSnackbar('error', 'Something went wrong! Unable to download schedule.', 5);
            console.log(error);
            return;
        }

        // Add timezone information to start and end times for events
        const icsString = value
            .replaceAll('DTSTART', 'DTSTART;TZID=America/Los_Angeles')
            .replaceAll('DTEND', 'DTEND;TZID=America/Los_Angeles');

        // Inject the VTIMEZONE section into the .ics file.
        const data = new Blob([icsString.replace('BEGIN:VEVENT', vTimeZoneSection)], {
            type: 'text/plain;charset=utf-8',
        });

        // Download the .ics file
        saveAs(data, 'schedule.ics');
        openSnackbar('success', 'Schedule downloaded!', 5);
    });
}
