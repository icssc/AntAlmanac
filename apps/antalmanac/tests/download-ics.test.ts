import { EventAttributes } from 'ics';
import type { Schedule } from '@packages/antalmanac-types';
import { describe, test, expect } from 'vitest';
import { getEventsFromCourses } from '$lib/download';
import { CalendarEvent } from '$components/Calendar/CourseCalendarEvent';

describe('download-ics', () => {
    test('converts schedule courses to events for the ics library', () => {
        const courses: CalendarEvent[] = [
            // CourseEvent
            {
                color: 'placeholderColor',
                start: new Date(2023, 9, 29, 1, 2),
                end: new Date(2023, 9, 29, 3, 4),
                title: 'placeholderDeptCode placeholderCourseNumber',
                locations: [{ building: 'placeholderLocation', room: 'placeholderRoom', days: 'MWF' }],
                showLocationInfo: true,
                finalExam: {
                    examStatus: 'SCHEDULED_FINAL',
                    dayOfWeek: 'Mon',
                    month: 2,
                    day: 3,
                    startTime: {
                        hour: 1,
                        minute: 2,
                    },
                    endTime: {
                        hour: 3,
                        minute: 4,
                    },
                    locations: [{ building: 'placeholderFinalLocation', room: 'placeholderFinalRoom' }],
                },
                courseTitle: 'placeholderCourseTitle',
                instructors: ['placeholderInstructor1', 'placeholderInstructor2'],
                isCustomEvent: false,
                sectionCode: 'placeholderSectionCode',
                sectionType: 'placeholderSectionType',
                term: '2023 Fall', // Cannot be a random placeholder; it has to be in `quarterStartDates` otherwise it'll be undefined
            },
            // CustomEvent
            {
                color: 'placeholderColor',
                start: new Date(2023, 9, 29, 1, 2),
                end: new Date(2023, 9, 29, 3, 4),
                title: 'placeholderCustomEventTitle',
                customEventID: 123,
                isCustomEvent: true,
                days: ['M', 'W', 'F'],
            },
        ];

        const result = getEventsFromCourses(courses);

        expect(result).toMatchSnapshot();
    });

    test('ics file has the correct contents', () => {
        /* TODO */
    });
});
