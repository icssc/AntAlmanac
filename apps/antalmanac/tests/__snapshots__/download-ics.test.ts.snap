// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`download-ics > converts schedule courses to events for the ics library 1`] = `
[
  {
    "description": "placeholderCourseTitle
Taught by placeholderInstructor1/placeholderInstructor2",
    "end": [
      2023,
      9,
      29,
      3,
      4,
    ],
    "endOutputType": "local",
    "location": "placeholderLocation placeholderRoom",
    "productId": "antalmanac/ics",
    "recurrenceRule": "FREQ=WEEKLY;BYDAY=FR,MO,WE;INTERVAL=1;COUNT=31",
    "start": [
      2023,
      9,
      29,
      1,
      2,
    ],
    "startOutputType": "local",
    "title": "placeholderDeptCode placeholderCourseNumber placeholderSectionType",
  },
  {
    "description": "Final Exam for placeholderSectionType placeholderCourseTitle",
    "end": [
      2023,
      3,
      3,
      3,
      4,
    ],
    "endOutputType": "local",
    "productId": "antalmanac/ics",
    "start": [
      2023,
      3,
      3,
      1,
      2,
    ],
    "startOutputType": "local",
    "title": "placeholderDeptCode placeholderCourseNumber Final Exam",
  },
  {
    "end": [
      2023,
      1,
      8,
      3,
      4,
    ],
    "endOutputType": "local",
    "productId": "antalmanac/ics",
    "recurrenceRule": "FREQ=WEEKLY;BYDAY=MO,WE,FR;INTERVAL=1;COUNT=30",
    "start": [
      2023,
      1,
      8,
      1,
      2,
    ],
    "startOutputType": "local",
    "title": "placeholderCustomEventTitle",
  },
]
`;
